sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memcpy.asciidoc

```diff
@@ -0,0 +1,161 @@
+= sycl_ext_oneapi_async_memcpy
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:dpcpp: pass:[DPC++]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2023-2023 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 7 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+This extension also depends on the following SYCL extensions:
+
+* link:https://github.com/intel/llvm/pull/9186/[sycl_ext_oneapi_barrier]
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+This extension defines
+`sycl::ext::oneapi::experimental::async_memcpy` free function to
+generalize and replace the current `sycl::async_work_group_copy`
+function. 
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_ASYNC_MEMCPY` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== `async_memcpy` function
+`sycl::ext::oneapi::experimental::async_memcpy` is a free function
+that  asynchronously copies a number of elements specified by
+`num_elements` of data of type `T` from the source pointer `src` to
+destination pointer `dest`. It also takes a barrier object of type
+`syclex::barrier` as an argument that can be used to wait on the
+completion of the memory copy.
+
+Permitted types for `T` are all scalar and vector types.
```

Comment:
I suspect that we may need some more restrictions here, or at least some non-normative notes to highlight potentially surprising behavior.

It's unsafe to use `memcpy` if `T` isn't trivially copyable.  This should probably be listed as a precondition of `memcpy_async` as well.

If `T` has a non-trivial copy constructor, developers calling `async_copy` would expect the copy constructor to execute.  That would (probably) prevent an implementation from using any hardware-acceleration for asynchronous copies.  I don't know whether it's better to limit `T` to trivially copyable types, or to add a non-normative note that many implementations will probably not execute asynchronously if `T` is not trivially copyable.

I'm leaning towards accepting all `T` and adding a note, because asynchrony already isn't guaranteed.

Reply:
I don't think we should limit this to scalar and vector. One can create a class and allocate it in local memory/registers where the hardware can use such accelerated async_memcpy for such class. 
I will add a precondition that `T` must be a trivially copyable type.